File: lib/data/services/sensor_service.dart
import 'dart:async';
import 'package:sensors_plus/sensors_plus.dart';
import '../models/sensor_data.dart';

class SensorService {
  /// Fixed output sampling rate for the recorded time series (Hz).
  /// The periodic timer samples the latest cached sensor reading at this
  /// rate, so the produced series is uniform regardless of the underlying
  /// platform sensor delivery rate. 100 Hz is a typical choice for gait.
  static const int samplingRateHz = 100;
  static const Duration _samplePeriod =
      Duration(microseconds: 1000000 ~/ samplingRateHz);

  StreamController<SensorData>? _controller;
  StreamSubscription<AccelerometerEvent>? _accelSubscription;
  StreamSubscription<GyroscopeEvent>? _gyroSubscription;
  Timer? _timer;

  // Store latest sensor readings
  List<double> _currentAccel = [0.0, 0.0, 0.0];
  List<double> _currentGyro = [0.0, 0.0, 0.0];

  Stream<SensorData> startRecording(String footSide) {
    _controller = StreamController<SensorData>.broadcast();

    // Listen to accelerometer
    _accelSubscription = accelerometerEventStream().listen((AccelerometerEvent event) {
      _currentAccel = [event.x, event.y, event.z];
    });

    // Listen to gyroscope
    _gyroSubscription = gyroscopeEventStream().listen((GyroscopeEvent event) {
      _currentGyro = [event.x, event.y, event.z];
    });

    // Emit data at a constant rate so the recorded series is uniform —
    // important for downstream models that expect a fixed sampling rate.
    _timer = Timer.periodic(_samplePeriod, (_) {
      if (_controller?.isClosed ?? true) return;

      final data = SensorData(
        timestamp: DateTime.now().millisecondsSinceEpoch,
        accX: _currentAccel[0],
        accY: _currentAccel[1],
        accZ: _currentAccel[2],
        gyrX: _currentGyro[0],
        gyrY: _currentGyro[1],
        gyrZ: _currentGyro[2],
        footSide: footSide,
      );
      _controller?.add(data);
    });

    return _controller!.stream;
  }

  Future<void> stopRecording() async {
    _timer?.cancel();
    await _accelSubscription?.cancel();
    await _gyroSubscription?.cancel();
    await _controller?.close();
    
    _timer = null;
    _accelSubscription = null;
    _gyroSubscription = null;
    _controller = null;
  }
}

File: src/data/services/sensor_service.js
import { SensorData } from '../models/sensor_data.js';

/**
 * SensorService — placeholder for sensors_plus integration.
 * In production this subscribes to device accelerometer + gyroscope streams.
 * For now generates dummy data to allow the full pipeline to run.
 */
export class SensorService {
  constructor() {
    this.isRecording = false;
    this._data = [];
    this._dummyInterval = null;
  }

  start(footSide) {
    this.isRecording = true;
    this._data = [];
    this._dummyInterval = setInterval(() => {
      this._data.push(this._generateSample(footSide));
    }, 20); // ~50 Hz dummy
  }

  stop() {
    this.isRecording = false;
    clearInterval(this._dummyInterval);
    return [...this._data];
  }

  getData() {
    return [...this._data];
  }

  _generateSample(footSide) {
    const t = Date.now();
    return new SensorData({
      timestamp: t,
      accX: (Math.random() - 0.5) * 2 + 0.1,
      accY: (Math.random() - 0.5) * 2 + 9.8,
      accZ: (Math.random() - 0.5) * 2 + 0.05,
      gyrX: (Math.random() - 0.5) * 0.5,
      gyrY: (Math.random() - 0.5) * 0.5,
      gyrZ: (Math.random() - 0.5) * 0.5,
      footSide,
    });
  }
}

export const sensorService = new SensorService();
